### ISSUE TITLE
Asking for driver installation every time and I/O error when trying to install driver

### ISSUE THREAD

**Issue by k0u5h1k** [ShadySquirrel/ViPER4Android#19]
If I install the driver from the app itself after installing the module, will it modify system anyway?

**Comment by ShadySquirrel** [comment]
As I answered on other repo, no, it should work out of the box unless your device is deviating from the norm. Please follow issue reporting guide amd provide necessary information

**Comment by ShadySquirrel** [comment]
Does that system by any chance have any other audio_effects.conf file? Pixels on android O have audio_effects.xml in /vendor/etc, for example, HTC devices have audio_effects_htc.conf etc etc.

Sepolicy is loaded nicely, drivers are incuded in audio configuration but for some reason V4A isn't seeing them, so it seems like there are more audio config files which probably have a higher priority than those two.

**Comment by k0u5h1k** [comment]
Only these files are present in /system/vendor/etc/  but there are no audio conf or xml file in /system/etc

![screenshot_20180118-175201](https://user-images.githubusercontent.com/35483317/35097976-77cde698-fc78-11e7-8398-15810dcb03de.png)
![screenshot_20180118-175216](https://user-images.githubusercontent.com/35483317/35097977-7840ee18-fc78-11e7-8200-4b78a92f4429.png)
![screenshot_20180118-175151](https://user-images.githubusercontent.com/35483317/35097978-788a35dc-fc78-11e7-93c0-9782aafd8aa0.png)

**Comment by ShadySquirrel** [comment]
Interesting... Well, thing is, looks like something is overriding audio configuration there so V4A app doesn't see it's library. I can't do much about that without device in my hands. If you can find that file somewhere, it'd be easy to fix, like it was with #18 as soon as we found out about that XML file.

**Comment by k0u5h1k** [comment]
If u can tell me how to do that, then I will surely help you..

**Comment by k0u5h1k** [comment]
Is there any chance that /vendor/etc/audio_policy.conf  is the file u are searching?

**Comment by ShadySquirrel** [comment]
I'd need any config file which has audio effects configuration - those mostly have audio_effects in it's name and are in JSON-like format like ones you've uploaded. audio_policy.conf is something v4a doesn't need to change, yet.

So if you can check all audio-related files from your system (from vendor, system/etc/, vendor/etc, I don't know how Xiaomi implemented stuff there) and find any trace of audio effects, please upload that file here for me to examine.

**Comment by k0u5h1k** [comment]
There are two audio_effects file.. 
/system_root/system/etc/audio_effects.conf
[system_root-system-etc-audio_effects.conf.gz](https://github.com/ShadySquirrel/ViPER4Android/files/1642865/system_root-system-etc-audio_effects.conf.gz)
/system_root/system/vendor/etc/audio_effects.conf
[system_root-system-vendor-etc-audio_effects.conf.gz](https://github.com/ShadySquirrel/ViPER4Android/files/1642869/system_root-system-vendor-etc-audio_effects.conf.gz)

**Comment by k0u5h1k** [comment]
Other than above 2 files, there are no files named audio_effects

**Comment by ShadySquirrel** [comment]
I'll check more stuff and get back to you

**Comment by k0u5h1k** [comment]
Ok thanks for looking into it

**Comment by k0u5h1k** [comment]
For ur kind infi, this device has dirac enabled.. if thats any help of u

**Comment by ShadySquirrel** [comment]
Yeah, but we have another issue - dirac is usually causing Processing: No error, not driver failure. If you think that disabling dirac would help, [try with this module](https://github.com/ShadySquirrel/ViPER4Android/issues/5#issuecomment-339637475) just remove current one and reboot first. It's basically a quick fix for Dirac, which is removing it's directories systemlessly, and will be included in next version.

**Comment by k0u5h1k** [comment]
Using ur modified module also had no luck for me...

**Comment by ShadySquirrel** [comment]
Yeah, that thing sometimes works sometimes not, that is why I blocked AudioFX and MusicFX differently...

Main problem with your device is still that unknown audio config file which has priority over two uploaded. Without adding V4A to that file, there can be little to no progress...

**Comment by k0u5h1k** [comment]
Any other possible names of that config file? I have searched for "audio_effects" strings under all the root directories.. only found 4 files that I have already uploaded..

**Comment by ShadySquirrel** [comment]
It should have audio_effects in name, extension, prefix or suffix may vary

**Comment by k0u5h1k** [comment]
Some people got it working with simply selinux permissive script and ur 5.1 module in the same phone and same rom but for most of the people its not working..

**Comment by ShadySquirrel** [comment]
Well other reason is probably dirac, but that can be circumvented. Thing is, from your log, there is no sepolicy breakage I've noticed that may stop v4a from loading the driver. I have no clue what Xiaomi did there

**Comment by k0u5h1k** [comment]
Well I have removed the diracaudioservice using the magisk way and still no improvement..

**Comment by ShadySquirrel** [comment]
That means it haven't found a driver. If issue was selinux, it would report a driver version, but status would stay at 'Abnormal'. 

I can't do much without device in my hands or without proper config files; and I'm not familiar with inner workings of anything Xiaomi so unless someone provides a device or needed files, I can't do much.

**Comment by k0u5h1k** [comment]
But the module is working with selinux permissive script for some people in the same device and the same stock rom without any other modification.. If I provide logs from one of them, can you determine the issue?

**Comment by ShadySquirrel** [comment]
I guess I could look (all configuration files, too), but I wouldn't hope much.

TBH until V4A is officially supporting 8.x, I can't do much.

**Comment by k0u5h1k** [comment]
I dont know how but after repeated try with uninstalling and installing modules, magisk itsepf finally got it working.. as of now, its working with magisk 15.2 and selinux permissive .. 
![screenshot_20180119-184645](https://user-images.githubusercontent.com/35483317/35152909-52c6df1c-fd4a-11e7-81c2-e17ba8c39552.png)

**Comment by k0u5h1k** [comment]
again after some reboots driver went missing.. please bro dont give up.. you can fix it.

**Comment by k0u5h1k** [comment]
Finally got the viper working by disabling secured boot protection.. pattern that appears before every boot.. that was causing all the problems.. anyway now it requires selinux permissive to get the module working.. now the driver version is showing but in abnormal status.. after using selinux permissive, its working fine... can you remove this limitation??

**Comment by ShadySquirrel** [comment]
Whoa, I'm not available for a day and you nearly fixed it :D

Now I need logcat (`adb logcat -b all`), with with both V4A during enforced and permissive, and your willingness to test sepolicy changes.

**Comment by k0u5h1k** [comment]
Ya bro spent a lot of time to test the possibility of working viper.. I am ready to test your changes.. here are the logs with selinux enforcing and permissive..
[log_enforcing.txt](https://github.com/ShadySquirrel/ViPER4Android/files/1648504/log_enforcing.txt)
[log_permissive.txt](https://github.com/ShadySquirrel/ViPER4Android/files/1648506/log_permissive.txt)

**Comment by ShadySquirrel** [comment]
I'm seeing similar audit output to one mentioned [here](https://github.com/ShadySquirrel/ViPER4Android/issues/18#issuecomment-358490661). Nearly missed it because it gets denied only once or twice? And log is full of V4A respawning.

I hope you know your way around, lol.

Like I've mentioned in that comment, I need you to edit `/magisk/viper4android/post-fs-data.sh` and add this line after last line in that file (the OnePlus fix):
`magiskpolicy --live "allow hal_audio_default hal_audio_default_tmpfs file { execute }"`

Save changes, reboot, start logging as soon as possible.

P.S. Keep Magisk mount or uninstall zip close, in case you aren't able to boot.

**Comment by k0u5h1k** [comment]
Its worked bro .. thanks a lot bro.. you rocked..

**Comment by ShadySquirrel** [comment]
You're welcome :)

I'll push this fix and update the module as soon as I get response if it's fixing issues on Pixel devices.

Thanks for testing!

I'm marking this as solved and closing this thread.
